gpiozero/devices.py

```diff
@@ -229,13 +189,100 @@ def __exit__(self, exc_type, exc_value, exc_tb):
 class Device(ValuesMixin, GPIOBase):
     """
     Represents a single device of any type; GPIO-based, SPI-based, I2C-based,
-    etc. This is the base class of the device hierarchy. It defines the
-    basic services applicable to all devices (specifically the :attr:`is_active`
+    etc. This is the base class of the device hierarchy. It defines the basic
+    services applicable to all devices (specifically the :attr:`is_active`
     property, the :attr:`value` property, and the :meth:`close` method).
     """
+    _pin_factory = None # instance of a Factory sub-class
+    _reservations = {} # maps pin addresses to lists of devices
```

Comment:
Ah, that's a subtle point:

1. If someone's evil enough to use multiple local factories the reservation system will still catch them
2. Someone may legitimately use multiple factories if they're, e.g. remote PiGPIOFactory instances. But still the reservation system has to remain centralized (it's not perfect, i.e. it won't defend against "localhost/GPIO12" vs "127.0.0.1/GPIO12" but you can't block *every* foot-gun combination ;)